### DISCUSSION TITLE
React compiler and useMemo deprecation

### DISCUSSION THREAD

**Discussion by krskrs** [preactjs/preact#4420] [General]
I have seen that React is introducing a new Compiler which would essentially make useMemo and React.memo useless. I don't know if this will eventually lead to the deprecation and removal of useMemo and React.memo, but even if they weren't deprecated, I believe that libraries would start stopping using them, as the aim of the compiler is just handling such optimizations automatically, under the hood, at compile time.
So, I'm worried that if libraries stop using such methods, using them in Preact would, at the minimum, bring to a lower performance?
What do you think?

**Comment by marvinhagemeister** [comment]
Preact users are generally gravitating more towards a signal based approach to managing state in apps. This eliminates the performance problems the React compiler is trying to address entirely as Signals ensure that state updates are decoupled from rendering and only trigger components to re-render when they need to. This means both `useMemo` and `React.memo` isn't needed with Signals and you don't even need to worry about dependency arguments.

The beauty about Signals is that they are not constrained to a particular framework or pattern and can be used anywhere. They don't require a complex machinery to make them work like the React compiler as they're plain JavaScript. I've used them in some personal non-preact apps and even backend-only apps a few times.

**Comment by krskrs** [comment]
Hi @marvinhagemeister , I get your point, and I totally agree with that, but maybe I haven't been clear about what was worrying me or I haven't fully understood your reply. I was focusing more on the usage of libraries made for the react ecosystem with Preact. For example, I'm using React-mui or React-router with preact with 0 issues, and I was wondering what will happen when (and if?) such libraries would stop using internally useMemo to totally rely on the new compiler. 
But whoops, 😄  as I'm writing this reply, I'm realizing that maybe my worry doesn't make any sense, as the compiled versions would still work on preact, and they only just won't use useMemo from Preact, but they would just be intrinsically optimized (i mean at build time, having been built with the react compilter).
Makes sense? 
PS I decided to post this reply anyway, because maybe someone else could got confused in the same way as I was, and could have the same doubt.

**Comment by rschristian** [comment]
> I don't know if this will eventually lead to the deprecation and removal of useMemo and React.memo

As far as I'm aware, the compiler generates output that uses `useMemo` and `React.memo`, and besides, with the back-compat, they're not going anywhere.

> I believe that libraries would start stopping using them, as the aim of the compiler is just handling such optimizations automatically, under the hood, at compile time.

No, the official recommendation is to ship pre-compiled libs to NPM: 

> yeah, we recommend library authors to test and ship precompiled versions
>
> we think compat shouldn't be too an issue as the compiled code only depends on one hook, but this is the kind of feedback we'd like from the wg

https://x.com/_gsathya/status/1792968851212169441
